pkg/executor/api.go

```diff
@@ -50,25 +50,59 @@ func (executor *Executor) getServiceForFunctionAPI(w http.ResponseWriter, r *htt
 	}
 
 	t := fn.Spec.InvokeStrategy.ExecutionStrategy.ExecutorType
-	et, exists := executor.executorTypes[t]
-	if !exists {
-		http.Error(w, fmt.Sprintf("Unknown executor type '%v'", t), http.StatusNotFound)
-		return
-	}
+	et := executor.executorTypes[t]
 
-	executor.logger.Debug(fmt.Sprintf("active instances: %v", et.GetTotalAvailable(fn)))
-	conncurrency := fn.Spec.Concurrency
-	if conncurrency == 0 {
-		// set to default conncurrency
-		conncurrency = 5
-		executor.logger.Debug(fmt.Sprintf("concurrency specified in function: %v", fn.Spec.Concurrency))
-		executor.logger.Debug("setting concurrency to 5")
-	}
-	if t == fv1.ExecutorTypePoolmgr && et.GetTotalAvailable(fn) >= conncurrency {
-		errMsg := fmt.Sprintf("max concurrency reached for %v. All %v instance are active", fn.ObjectMeta.Name, conncurrency)
-		executor.logger.Error("error occurred", zap.String("error", errMsg))
-		http.Error(w, errMsg, http.StatusTooManyRequests)
-		return
+	// Check function -> svc cache
+	executor.logger.Debug("checking for cached function service",
+		zap.String("function_name", fn.ObjectMeta.Name),
+		zap.String("function_namespace", fn.ObjectMeta.Namespace))
+	if t == fv1.ExecutorTypePoolmgr {
+		concurrency := fn.Spec.Concurrency
+		if concurrency == 0 {
+			concurrency = 5
+		}
+		requestsPerpod := fn.Spec.RequestsPerPod
+		if requestsPerpod == 0 {
+			requestsPerpod = 1
+		}
+		fsvc, active, err := et.GetFuncSvcFromPoolCache(fn, requestsPerpod)
+		if err == nil {
+			// if pod is already serving request than it already exist else validated
+			executor.logger.Debug("from cache", zap.Int("active", active))
+			if active > 1 || et.IsValid(fsvc) {
+				// Cached, return svc address
+				executor.logger.Debug("served from cache", zap.String("name", fsvc.Name), zap.String("address", fsvc.Address))
+				executor.writeResponse(w, fsvc.Address, fn.ObjectMeta.Name)
+				return
+			}
+			executor.logger.Debug("deleting cache entry for invalid address",
+				zap.String("function_name", fn.ObjectMeta.Name),
+				zap.String("function_namespace", fn.ObjectMeta.Namespace),
+				zap.String("address", fsvc.Address))
+			et.DeleteFuncSvcFromCache(fsvc)
+			active--
```

Comment:
Should this not go in an else block for the `isValid` condition? Otherwise, this will be executed even if a service `isValid` by default and when active <= 1?

Reply:
In '||' condition if the first condition is evaluated to true, subsequent conditions aren't evaluated.